pkg/metrics/instance/group_manager.go

```diff
@@ -30,11 +37,152 @@ type GroupManager struct {
+}
+
+// ApplyConfigs is used to batch configurations for performance instead of the singular ApplyConfig
+func (m *GroupManager) ApplyConfigs(configs []Config) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	return m.applyConfigs(configs, false)
+}
+
+// applyConfigs is used to take the currently running set of configs, plus the new ones. (update/append configs).
+// isRollback determines whether we are rolling back any changes and if so avoid recursion in calling rollback again
+func (m *GroupManager) applyConfigs(configs []Config, isRollback bool) error {
+	if len(configs) == 0 {
+		return nil
+	}
+	failed := make([]BatchFailure, 0)
+
+	// This will form our master set of configurations to apply by their group, this includes any that are currently
+	// running but not in the configs parameter.
+	groupsInConfigs := make(map[string]groupedConfigs)
+
+	// Combined group of current and new configs
+	combinedConfigs := m.getCombinedConfigs(configs)
+
+	oldConfigs := make([]Config, 0)
+	// In case of an error in applying the new configs, we need to grab the old config for rollback
+	copy(oldConfigs, m.activeConfigs)
+	activeConfigs := make([]Config, 0)
+	groupLookup := make(map[string]string)
+	// Iterate through the combined configurations
+	for _, c := range combinedConfigs {
+		groupName, err := hashConfig(c)
+		if err != nil {
+			failed = append(failed, BatchFailure{
+				Err:    err,
+				Config: c,
+			})
+			level.Error(m.log).Log("err", fmt.Sprintf("failed to get group name for config %s: %s", c.Name, err))
+			continue
+		}
+		if _, exists := groupsInConfigs[groupName]; !exists {
+			groupsInConfigs[groupName] = make(groupedConfigs)
+		}
+		groupsInConfigs[groupName][c.Name] = c
+		groupLookup[c.Name] = groupName
+		activeConfigs = append(activeConfigs, c)
+	}
+	// If we have group that no longer exist then we need to delete them
+	for groupName := range m.groups {
+		if _, exists := groupsInConfigs[groupName]; !exists {
+			if err := m.inner.DeleteConfig(groupName); err != nil {
+				level.Error(m.log).Log("err", fmt.Sprintf("failed to delete group named  %s with error %s", groupName, err))
+			}
+		}
+	}
+	groups := make(map[string]groupedConfigs)
+	mergedHashes := make(map[string]int)
+	// Now that we have grouped all the new and existing configurations we can apply them
+	for groupName, grouped := range groupsInConfigs {
+		mergedConfig, err := groupConfigs(groupName, grouped)
+		if err != nil {
+			level.Error(m.log).Log("err", fmt.Sprintf("failed to group configs with groupname  %s with error %s", groupName, err))
+		}
+		// If we have the exact same config no need to reload it
+		newHash, err := createMergedConfigHash(mergedConfig)
+		if err != nil {
+			level.Error(m.log).Log("err", fmt.Sprintf("failed to create merged hash named  %s with error %s", groupName, err))
+			continue
+		}
+		groups[groupName] = grouped
+		mergedHashes[newHash] = 0
+		if _, exists := m.mergedConfigHashes[newHash]; exists {
+			continue
+		}
```

Comment:
This would cause unexpected issues if two groups happen to have the same hash. We should find a way to associate the hash with the group.

Reply:
Then we also have that concern with the hashConfig function also?

Reply:
Yeah, I think you're right. In that case you do want hash collisions at least at the group level, but it's still possible that two incompatible configs have the same group hash.